packages/punchcard/lib/appsync/vtl-object.ts

```diff
@@ -0,0 +1,208 @@
+export class VInteger extends VObject<IntegerShape> {}
+export class VFloat extends VObject<NumberShape> {}
+export class VNothing extends VObject<NothingShape> {}
+export class VNever extends VObject<NeverShape> {}
+export class VBinary extends VObject<BinaryShape> {}
+
+export class VBool extends VObject<BoolShape> {
+  public static not(a: VBool): VBool {
+    return new VBool(bool, VObject.exprOf(a).prepend('!'));
+  }
+}
+
+export class VString extends VObject<StringShape> {
+  public toUpperCase(): VString {
+    return new VString(VObject.typeOf(this), VObject.exprOf(this).dot('toUpperCase()'));
+  }
+
+  public isNotEmpty(): VBool {
+    return VBool.not(this.isEmpty());
+  }
+
+  public isEmpty(): VBool {
+    return new VBool(bool, VObject.exprOf(this).dot('isEmpty()'));
+  }
+
+  public size(): VInteger {
+    return new VInteger(integer, VObject.exprOf(this).dot('size()'));
+  }
+}
+
+export class VTimestamp extends VObject<TimestampShape> {}
+
+export class VList<T extends VObject = VObject> extends VObject<ArrayShape<VObject.TypeOf<T>>> {
+  constructor(shape: ArrayShape<VObject.TypeOf<T>>, expression: VExpression) {
+    super(shape, expression);
+  }
+}
+
+
+export class VSet<T extends VObject = VObject> extends VObject<SetShape<VObject.TypeOf<T>>> {
+  constructor(shape: SetShape<VObject.TypeOf<T>>, expression: VExpression) {
+    super(shape, expression);
+  }
+}
+
+export class VRecord<M extends VRecord.Members = {}> extends VObject<RecordShape<{
+  [m in keyof M]: M[m][typeof type];
+}>> {
+  constructor(type: RecordShape<any, any>, expr: VExpression) {
+    super(type, expr);
+    for (const [name, shape] of Object.entries(type.Members) as [string, Shape][]) {
+      (this as any)[name] = VObject.of(shape, expr.dot(name));
+    }
+  }
+}
+export namespace VRecord {
+  export type GetMembers<R extends VRecord> = R extends VRecord<infer M> ? M : any;
+  export interface Members {
+    [m: string]: VObject;
+  }
+  export type Class<T extends VRecord = any> = (new(members: VRecord.GetMembers<T>) => T);
+}
+
+export class Visitor implements ShapeVisitor<VObject, VExpression> {
+  public static defaultInstance = new Visitor();
+
+  public functionShape(shape: FunctionShape<FunctionArgs, Shape>): VObject<Shape> {
+    throw new Error("Method not implemented.");
+  }
+  public neverShape(shape: NeverShape, context: VExpression): VObject<Shape> {
+    return new VNever(shape, context);
+  }
+  public arrayShape(shape: ArrayShape<any>, expr: VExpression): VList {
+    return new VList(shape, expr);
+  }
+  public binaryShape(shape: BinaryShape, expr: VExpression): VBinary {
+    return new VBinary(shape, expr);
+  }
+  public boolShape(shape: BoolShape, expr: VExpression): VBool {
+    return new VBool(shape, expr);
+  }
+  public recordShape(shape: RecordShape<any>, expr: VExpression): VRecord {
+    return new VRecord(shape, expr);
+  }
+  public dynamicShape(shape: DynamicShape<any>, expr: VExpression): VAny | VUnknown {
```

Comment:
Similar to the comment above... will this shape ever be encountered by the visitor?

Reply:
While these types are not valid GraphQL schema types, the, we still need to interact with services like DynamoDB, SQS, SNS, Lambda which all have "VTL Representations".  For example, I treat sets as an array in JSON, but have to convert to a specific set type when serializing to/from DynamoDB.

We should be able to make use of the resolver utilities to implement some behavior for any. We could, for example, look for a "type tag" and cast to a well known type, or implement some generic serialization logic within VTL.